rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -836,148 +823,282 @@ rmw_ret_t configure_qos_for_security(
 #endif
 }
 
-
-extern "C" rmw_node_t * rmw_create_node(
-  rmw_context_t * context, const char * name,
-  const char * namespace_, size_t domain_id
-#if !RMW_VERSION_GTE(0, 8, 2)
-  , const rmw_node_security_options_t * security_options
-#endif
-#if RMW_VERSION_GTE(0, 8, 1)
-  , bool localhost_only
-#endif
-)
+extern "C" rmw_ret_t rmw_init(const rmw_init_options_t * options, rmw_context_t * context)
 {
-#if !RMW_VERSION_GTE(0, 8, 2)
+  rmw_ret_t ret;
+
+  static_cast<void>(options);
   static_cast<void>(context);
-#endif
-  RET_NULL_X(name, return nullptr);
-  RET_NULL_X(namespace_, return nullptr);
-#if MULTIDOMAIN
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(options, RMW_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(context, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    options,
+    options->implementation_identifier,
+    eclipse_cyclonedds_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+
   /* domain_id = UINT32_MAX = Cyclone DDS' "default domain id".*/
-  if (domain_id >= UINT32_MAX) {
+  if (options->domain_id >= UINT32_MAX) {
     RCUTILS_LOG_ERROR_NAMED(
       "rmw_cyclonedds_cpp", "rmw_create_node: domain id out of range");
-    return nullptr;
+    return RMW_RET_INVALID_ARGUMENT;
   }
-  const dds_domainid_t did = static_cast<dds_domainid_t>(domain_id);
-#else
-  static_cast<void>(domain_id);
-  const dds_domainid_t did = DDS_DOMAIN_DEFAULT;
-#endif
-#if !RMW_VERSION_GTE(0, 8, 2)
-  RCUTILS_CHECK_ARGUMENT_FOR_NULL(security_options, nullptr);
-#else
-  rmw_security_options_t * security_options;
-  RCUTILS_CHECK_ARGUMENT_FOR_NULL(context, nullptr);
-  security_options = &context->options.security_options;
-#endif
-  rmw_ret_t ret;
-  int dummy_validation_result;
-  size_t dummy_invalid_index;
-  if ((ret =
-    rmw_validate_node_name(name, &dummy_validation_result, &dummy_invalid_index)) != RMW_RET_OK)
-  {
-    return nullptr;
+  const dds_domainid_t domain_id = static_cast<dds_domainid_t>(options->domain_id);
+
+  context->instance_id = options->instance_id;
+  context->implementation_identifier = eclipse_cyclonedds_identifier;
+  context->impl = nullptr;
+
+  if ((ret = rmw_init_options_copy(options, &context->options)) != RMW_RET_OK) {
+    return ret;
+  }
+
+  std::unique_ptr<rmw_context_impl_t> impl(new(std::nothrow) rmw_context_impl_t());
+  if (impl == nullptr) {
+    return RMW_RET_BAD_ALLOC;
   }
 
-#if SUPPORT_LOCALHOST
   /* Take domains_lock and hold it until after the participant creation succeeded or
      failed: otherwise there is a race with rmw_destroy_node deleting the last participant
      and tearing down the domain for versions of Cyclone that implement the original
      version of dds_create_domain that doesn't return a handle.  */
-  std::lock_guard<std::mutex> lock(gcdds.domains_lock);
-  if (!check_create_domain_locked(did, localhost_only)) {
-    return nullptr;
+  if (!check_create_domain(domain_id, options->localhost_only)) {
```

Comment:
In `rmw_fastrtps`, a `Participant` is created when the first `Node` is created within a `Context`.
This is also a valid option, and shouldn't create cross-vendor problems.

Reply:
I have no idea when the context gets initialized relative to when the first node is created. If the two always go hand in hand, then it is not worth changing. But if applications often create the context much earlier (say, before checking their command-line arguments), it would be better to do it the same way as in `rmw_fastrtps`. I'm happy to go with this for now.